src/client/gemini_client.rs

```diff
@@ -0,0 +1,287 @@
+use bytes::Bytes;
+use nu_protocol::ShellError;
+use reqwest::Response;
+use serde::{Deserialize, Serialize};
+
+pub struct GeminiClient {
+    api_key: String,
+    max_tokens: usize,
+}
+
+// While Gemini does not have a per request limit, the per minute token limit is used here
+const MAX_FREE_TIER_TOKENS: usize = 1000000;
+
+// According to the Gemini API docs: A token is equivalent to about 4 characters for Gemini models
+const CHARS_PER_TOKEN: usize = 4;
+
+const MAX_EMBEDDING_DIMENSION: u32 = 768;
+
+// At most 100 requests can be in one batch
+const MAX_BATCH_SIZE: usize = 100;
+
+impl GeminiClient {
+    pub fn new(api_key: String, max_tokens: Option<usize>) -> Self {
+        let max_tokens = match max_tokens {
+            Some(mt) => mt,
+            None => MAX_FREE_TIER_TOKENS,
+        };
+
+        Self {
+            api_key,
+            max_tokens,
+        }
+    }
+
+    pub fn batch_chunks(&self, chunks: Vec<String>) -> Vec<Vec<String>> {
+        let mut tokens = 0;
+        let mut batch = vec![];
+        let mut batches = vec![];
+        for chunk in chunks {
+            tokens = tokens + chunk.chars().count() / CHARS_PER_TOKEN;
+
+            if tokens >= self.max_tokens || batch.len() == MAX_BATCH_SIZE {
+                batches.push(batch);
+                batch = vec![chunk.clone()];
+                tokens = chunk.chars().count();
+            } else {
+                batch.push(chunk.to_string());
+            }
+        }
+
+        batches.push(batch);
+        batches
+    }
+
+    pub async fn embed(&self, batch: &Vec<String>, dim: u32) -> Result<Vec<Vec<f32>>, ShellError> {
+        if dim > MAX_EMBEDDING_DIMENSION {
+            return Err(ShellError::GenericError {
+                error: format!(
+                    "Maximum embedding size is {:?} for Gemini",
+                    MAX_EMBEDDING_DIMENSION
+                ),
+                msg: "".to_string(),
+                span: None,
+                help: None,
+                inner: vec![],
+            });
+        }
+
+        let url = format!(
+            "https://generativelanguage.googleapis.com/v1/models/text-embedding-004:batchEmbedContents?key={}",
+            self.api_key
+        );
+
+        let mut batch_request = EmbeddingBatchRequest { requests: vec![] };
```

Comment:
I think this would be nicer as EmbeddingBatchRequest::new() where it just automatically sets requests to vec![] inside of new?

Reply:
Done!